Change get_url module to always use GET method, even in check mode

dantheta opened the pull request:
##### SUMMARY

Makes the get_url module check mode behaviour more realistic by always downloading and comparing the files, instead of using HEAD requests in check mode.

Fixes #61369

When remote checksum files are being used with check mode, ensures that the checksum file content is downloaded so that hash verification can take place and the check mode results will be most realistic.

With this patch applied, get_url's behaviour in check mode matches non-check mode.

Before appliying this patch, get_url in check mode would always report `changed: yes` for an existing, identical destination file.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

get_url

##### ADDITIONAL INFORMATION

Test playbook
```
- hosts: localhost
  tasks:
  - get_url:
      url: https://example.com/index.html
      dest: /tmp/example_index.html
      force: yes
```

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

# Before patch applied, with no existing destination file:

# Run in check mode

$ ansible-playbook  -C test.yml -v

PLAY [localhost] ***************************************************************

TASK [get_url] *****************************************************************
changed: [localhost] => {"changed": true, "checksum_dest": null, "checksum_src": "da39a3ee5e6b4b0d3255bfef95601890afd80709", "dest": "/tmp/example_index.html", "elapsed": 0, "msg": "OK (648 bytes)", "src": "/home/daniel/.ansible/tmp/ansible-tmp-1587584953.92-4446-80049134333569/tmp9ZIUfh", "url": "http://www.example.com/index.html"}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 

# now not in check mode:

$ ansible-playbook   test.yml -v

PLAY [localhost] ***************************************************************

TASK [get_url] *****************************************************************
changed: [localhost] => {"changed": true, "checksum_dest": null, "checksum_src": "4a3ce8ee11e091dd7923f4d8c6e5b5e41ec7c047", "dest": "/tmp/example_index.html", "elapsed": 0, "gid": 1000, "group": "daniel", "md5sum": "84238dfc8092e5d9c0dac8ef93371a07", "mode": "0644", "msg": "OK (1256 bytes)", "owner": "daniel", "size": 1256, "src": "/home/daniel/.ansible/tmp/ansible-tmp-1587585032.09-4491-42949354018662/tmpETi0BB", "state": "file", "status_code": 200, "uid": 1000, "url": "http://www.example.com/index.html"}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 

# dest file now exists, run in check mode:

$ ansible-playbook  -C test.yml -v

PLAY [localhost] ***************************************************************

TASK [get_url] *****************************************************************
changed: [localhost] => {"changed": true, "checksum_dest": "4a3ce8ee11e091dd7923f4d8c6e5b5e41ec7c047", "checksum_src": "da39a3ee5e6b4b0d3255bfef95601890afd80709", "dest": "/tmp/example_index.html", "elapsed": 0, "gid": 1000, "group": "daniel", "mode": "0644", "msg": "OK (648 bytes)", "owner": "daniel", "size": 1256, "src": "/home/daniel/.ansible/tmp/ansible-tmp-1587585082.47-4551-224605562482286/tmpIEG_48", "state": "file", "uid": 1000, "url": "http://www.example.com/index.html"}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 

# this is less accurate than it could be.  The downloaded temporary file is empty, because a HEAD request was issued.  The value for checksum_src is that of an empty string.
A task that used a remote checksum file would throw an error here, because the downloaded checksum file is also empty.

# Re-running the test after applying the patch:

$ ansible-playbook  -C test.yml -v

PLAY [localhost] ***************************************************************

TASK [get_url] *****************************************************************
ok: [localhost] => {"changed": false, "checksum_dest": "4a3ce8ee11e091dd7923f4d8c6e5b5e41ec7c047", "checksum_src": "4a3ce8ee11e091dd7923f4d8c6e5b5e41ec7c047", "dest": "/tmp/example_index.html", "elapsed": 0, "gid": 1000, "group": "daniel", "mode": "0644", "msg": "OK (1256 bytes)", "owner": "daniel", "size": 1256, "src": "/home/daniel/.ansible/tmp/ansible-tmp-1587585203.3-4812-99300720681182/tmpYGJKR8", "state": "file", "uid": 1000, "url": "http://www.example.com/index.html"}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 

# src and dest checksums match, task result is "ok" instead of "changed".


```

felixfontein commented on lib/ansible/modules/net_tools/basics/get_url.py:
```diff
@@ -26,8 +26,8 @@
        your proxy environment for both protocols is correct.
-     - From Ansible 2.4 when run with C(--check), it will do a HEAD request to validate the URL but
-       will not download the entire file or verify it against hashes.
+     - From Ansible 2.4 when run with C(--check), it will do a GET request to validate the URL and
```
That's obviously wrong, since older versions like 2.4 won't change.

dantheta replied:
You're quite right, my apologies.  I'm pushing a new version which adds a new comment, and I'll add a changelog fragment too.

felixfontein reviewed (commented):
This is a breaking change, so it needs to be properly documented as changed from 2.10 on (i.e. it also needs a porting guide entry next to a changelog fragment with `minor_changes` section).

dantheta committed:
Change get_url module to always use GET method, even in check mode; fixes ansible/ansible#61369

Makes check mode more realistic by always downloading and comparing the files.

When remote checksum files are being used, ensures that the checksum file
is downloaded when in check mode.

dantheta committed:
update module description with new GET behaviour

dantheta committed:
add changelog fragment for behaviour change

dantheta committed:
Add porting guide doc for get_url change

dantheta committed:
whitespace fix

Akasurde commented:
superseded by https://github.com/ansible/ansible/pull/66700

@dantheta Thanks for the contribution.
